### ISSUE TITLE
Include order customization

### ISSUE THREAD

**Issue by ledif** [spack/spack#4115]
I'm trying to create a package for a project with custom Makefiles. In the project's Makefiles, the order of includes is important, as the project has files that are meant to override files in the other includes. I have created a project that mimics this behavior at https://github.com/ledif/cpp-spack-order.

For example, the project uses Boost as a dependency and includes files in `extern/include/boost/` that are meant to take precedence over files in the main Boost library. The way this is handled in the custom Makefile is placing the includes for the custom files first:

```
$(CXX) -std=c++11 -I./extern/include -I./tools/libstdc++/4.9.3/ -I$BOOST_ROOT/include main.cpp -o main
```

In this way, the project's custom files will be included first instead of the original Boost files.

However, looking at spack's compiler wrapper `cc`, it appears that all includes from dependencies are prepended:
```bash
# Read spack dependencies from the path environment variable
IFS=':' read -ra deps <<< "$SPACK_DEPENDENCIES"
for dep in "${deps[@]}"; do
    # Prepend include directories
    if [[ -d $dep/include ]]; then
        if [[ $mode == cpp || $mode == cc || $mode == as || $mode == ccld ]]; then
            args=("-I$dep/include" "${args[@]}")
        fi
    fi 
```
Is there a way to customize this behavior to place all dependencies' includes at the end instead of the beginning? If not, is there a standard way to address this problem?

#2374 also seems to mention a similar issue.

**Comment by davydden** [comment]
> the project uses Boost as a dependency and includes files in extern/include/boost/ that are meant to take precedence over files in the main Boost library.

i wonder why anyone would be doing this? If a package needs to build against some version of boost -- it should just take it as-is without trying to hack some modified headers. What happens when someone wants to link against this package and also directly against boost? Which headers should be used? 

> is there a standard way to address this problem?

IMO this is just a very controversial design and Spack can't do much (and should not!) in such circumstances. I would contact developers and ask them not to hack custom headers.

**Comment by ledif** [comment]
Changing the package to not rely on custom headers is not really a possibility for me.

However, it might be reasonable to patch Boost itself when the package is installed and use the patched version as a dependency instead. Is it possible to provide a patch for a package's dependencies?

**Comment by adamjstewart** [comment]
You would have to add the patch to boost. Maybe make a variant called `foobar-patch` that adds a patch required by the foobar package. Then your foobar package can depend on `boost+foobar-patch`. Not pretty, but it will work.

**Comment by ledif** [comment]
Thanks for the responses @adamjstewart and @davydden. I'll look into it.

**Comment by alalazo** [comment]
@ledif Any update?

**Comment by davydden** [comment]
@ledif with current state of Spack you can add patches to boost form downstream packages which use it. This way you will be able to build patched `boost` to be used with your package.

**Comment by ledif** [comment]
I was able to successfully build my package based on @adamjstewart's response.

I created a separate `boost-foobar` package that installed the patched Boost files from the `foobar` package into the `boost-foobar` prefix. By adding `boost-foobar` as a dependency to `foobar`, I was able to include the patched Boost includes before the upstream Boost includes.

@davydden: I found the section in the documentation related to [dependency patching](http://spack.readthedocs.io/en/latest/packaging_guide.html#dependency-dependency-patching) and it seems like a cleaner way to do this. The solution that I described works for me as of right now, but I'll take a look at using downstream patching as well.

In either case, this problem appears to be solved so I will close this issue.
